handler/src/test/java/io/netty/handler/ssl/ocsp/OcspTest.java

```diff
@@ -0,0 +1,471 @@
+            @Override
+            public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) throws Exception {
+                try {
+                    clientCauseRef[0] = cause;
+                } finally {
+                    latch.countDown();
+                }
+            }
+        };
+
+        ServerOcspContext serverOcspContext = new ServerOcspContext() {
+            @Override
+            public boolean optional() {
+                return false;
+            }
+
+            @Override
+            public byte[] staple() throws Exception {
+                throw new IllegalStateException();
+            }
+        };
+
+        ClientOcspContext clientOcspContext = new TestClientOcspContext(true);
+
+        handshake(latch, serverOcspContext, clientOcspContext, serverHandler, clientHandler);
+
+        // If the client rejects the staple then Client and Server will fail with
+        // Exceptions which contain these respective messages.
+        String clientMessage = clientCauseRef[0].getMessage();
+        assertTrue(clientMessage, clientMessage.contains("tlsv1 alert internal error"));
+
+        String serverMessage = serverCauseRef[0].getMessage();
+        assertTrue(serverMessage, serverMessage.contains("clienthello tlsext"));
+    }
+
+    private static void handshake(CountDownLatch latch, ServerOcspContext serverOcspContext,
+            ClientOcspContext clientOcspContext, ChannelHandler serverHandler,
+            ChannelHandler clientHandler) throws Exception {
+
+        SelfSignedCertificate ssc = new SelfSignedCertificate();
+        try {
+            SslContext serverSslContext = SslContextBuilder.forServer(ssc.certificate(), ssc.privateKey())
+                    .sslProvider(SslProvider.OPENSSL)
+                    .build();
+            ocspfySslContext(serverSslContext, serverOcspContext);
+
+            SslContext clientSslContext = SslContextBuilder.forClient()
+                    .sslProvider(SslProvider.OPENSSL)
+                    .trustManager(InsecureTrustManagerFactory.INSTANCE)
+                    .build();
+            ocspfySslContext(clientSslContext, clientOcspContext);
+
+            EventLoopGroup group = new DefaultEventLoopGroup();
+            try {
+                LocalAddress address = new LocalAddress("handshake-" + Math.random());
+                Channel server = newServer(group, address, serverSslContext, serverHandler);
+                Channel client = newClient(group, address, clientSslContext, clientHandler);
+                try {
+                    if (!latch.await(10L, TimeUnit.SECONDS)) {
+                        throw new IllegalStateException("Something went wrong.");
+                    }
+                } finally {
+                    client.close();
+                    server.close();
+                }
+            } finally {
+                group.shutdownGracefully();
```

Comment:
consider specifying the timeout values here so the unit tests don't have to wait for the quite period etc...